### PR TITLE
[plugin-mobile-app] Add possibility to disable screenshots downscale

### LINKED ISSUE #2195: [visual-testing] Add possibility to disable screenshots downscale
Currently, we downscale captured baselines using DPR. In order to perform visual checks without quality loss, we need to implement a way to disable downscaling. Tradeoffs are - visual testing precision vs baselines size. 
The feature should be togglable. Downscaling is enabled by the default

### PR DESCRIPTION
**Pull request by EDbarvinsky**
Closes #2195

### CODE REVIEW

**Comment by EDbarvinsky** [comment]
There is several things about this PR.
The first one is that for web application visual check  if dpr != 1 then it is not working correctly and current wrapping for cords provider does not fix that.
So there is a question should be documents be updated for web module?

Another one will we set up integration tests execution for mobile devices in case if we use original image instead ?

**Comment by EDbarvinsky** [comment]
Issue for web app is added https://github.com/vividus-framework/vividus/issues/2334

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractAshotFactory.java [hunk 1]
```diff
@@ -80,7 +80,7 @@ protected ShootingStrategy getBaseShootingStrategy()
         return ShootingStrategies.scaling((float) this.getDpr());
     }
 
-    protected abstract double getDpr();
+    public abstract double getDpr();
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractAshotFactory.java:83]: why is this change needed?

File: vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/DprAwareCoordsProvider.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    @Override
+    public Coords ofElement(WebDriver driver, WebElement element)
+    {
+        Coords coords = super.ofElement(driver, element);
+        return scale(coords);
+    }
+
+    private Coords scale(Coords coords)
+    {
+        int dpr = downscale ? 1 : (int) super.getMobileAppWebDriverManager().getDpr();
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/DprAwareCoordsProvider.java:47]: is `dpr` always integer?

File: vividus-plugin-mobile-app/src/main/resources/spring.xml [hunk 3]
```diff
@@ -92,19 +92,24 @@
 
     <bean class="org.vividus.ui.mobile.action.MobileAppElementActions" />
 
-    <bean class="org.vividus.selenium.mobileapp.screenshot.MobileAppAshotFactory" parent="abstractAshotFactory">
+    <bean id ="ashotFactory" class="org.vividus.selenium.mobileapp.screenshot.MobileAppAshotFactory" parent="abstractAshotFactory">
```
Comment (EDbarvinsky) [inline vividus-plugin-mobile-app/src/main/resources/spring.xml:95]: to remove

File: vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/MobileAppAshotFactory.java [hunk 4]
```diff
@@ -53,13 +55,27 @@ public AShot create(Optional<ScreenshotConfiguration> screenshotConfiguration)
         }).get();
         ShootingStrategy strategy = getStrategyBy(ashotConfig.getShootingStrategy().get())
             .getDecoratedShootingStrategy(getBaseShootingStrategy());
-        strategy = configureNativePartialsToCut(mobileAppWebDriverManager.getStatusBarSize(), ashotConfig, strategy);
+        int statusBarSize = downscale
+            ? mobileAppWebDriverManager.getStatusBarSize()
+            : mobileAppWebDriverManager.getStatusBarSize() * (int) getDpr();
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/MobileAppAshotFactory.java:60]:
```suggestion
        int statusBarSize = mobileAppWebDriverManager.getStatusBarSize()
        if (downscale)
        {
            statusBarSize = statusBarSize * (int) getDpr();
        }
```

File: vividus-plugin-mobile-app/src/main/resources/spring.xml [hunk 5]
```diff
@@ -92,19 +92,24 @@
...
                 <constructor-arg value="mobile.screenshot.strategy." />
                 <constructor-arg value="org.vividus.selenium.screenshot.ScreenshotConfiguration"/>
             </bean>
         </property>
         <property name="shootingStrategy" value="${mobile.screenshot.strategy}" />
+        <property name="downscale" value="${mobile.screenshot.downscale}" />
     </bean>
 
-    <bean id="mobileAppWebDriverManager" class="org.vividus.selenium.mobileapp.MobileAppWebDriverManager" parent="genericWebDriverManager"/>
+    <bean id="mobileAppWebDriverManager" class="org.vividus.selenium.mobileapp.MobileAppWebDriverManager" parent="genericWebDriverManager" />
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/resources/spring.xml:106]: noisy change

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-mobile-app/src/main/resources/spring.xml [hunk 6]
```diff
@@ -100,11 +100,14 @@
...
 
-    <bean id="coordsProvider" class="org.vividus.selenium.mobileapp.screenshot.NativeHeaderAwareCoordsProvider" />
+    <bean id="coordsProvider" class="org.vividus.selenium.mobileapp.screenshot.DprAwareCoordsProvider">
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/resources/spring.xml:108]: it seems `NativeHeaderAwareCoordsProvider` is not going to be used explicitly, so `DprAwareCoordsProvider` and `NativeHeaderAwareCoordsProvider` can be merged into a new `MobileAppCoordsProvider`

File: docs/modules/plugins/pages/plugin-visual.adoc [hunk 7]
```diff
@@ -209,6 +209,25 @@ image::example.png[Difference between native/web footer/header image]
...
+
+|`mobile.screenshot.downscale`
+|`boolean`
+|`true`
+|Downscale checkpoint image according to a device DPR value
+|===
+
+Please see the image to get a clue about difference between original and downscaled images.
+
+image::origininal_and_downscaled_checkpoints.png[Difference between original and dpr downscaled image]
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-visual.adoc:229]: add more details around the image: what is the original image, what is the downscaled image, what is DPR (please bear in mind it's not obvious for users)

File: vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/MobileAppCoordsProvider.java [hunk 8]
```diff
@@ -49,4 +54,19 @@ protected Coords getCoords(WebElement element)
...
+        return mobileAppWebDriverManager;
+    }
+
+    private Coords adjustToDpr(Coords coords)
+    {
+        double dpr = getMobileAppWebDriverManager().getDpr();
+        coords.x = CoordsUtil.multiply(coords.x, dpr);
+        coords.y = CoordsUtil.multiply(coords.y, dpr);
+        coords.width = CoordsUtil.multiply(coords.width, dpr);
+        coords.height = CoordsUtil.multiply(coords.height, dpr);
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/MobileAppCoordsProvider.java:69]:
```suggestion
        return CoordsUtil.scale(coords, dpr);
```

File: vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/util/CoordsUtil.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+public final class CoordsUtil
+{
+    private CoordsUtil()
+    {
+    }
+
+    public static int multiply(int value, double dpr)
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/util/CoordsUtil.java:28]:
```suggestion
    public static int scale(int value, double dpr)
```

File: vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/MobileAppCoordsProvider.java [hunk 10]
```diff
@@ -49,4 +54,19 @@ protected Coords getCoords(WebElement element)
         return new Coords(coords.x, coords.y - mobileAppWebDriverManager.getStatusBarSize(), coords.width,
                 coords.height);
     }
+
+    public MobileAppWebDriverManager getMobileAppWebDriverManager()
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/MobileAppCoordsProvider.java:58]: is it used?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/util/CoordsUtil.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    private CoordsUtil()
+    {
+    }
+
+    public static Coords scale(Coords coords, double dpr)
+    {
+        coords.x = CoordsUtil.scale(coords.x, dpr);
+        coords.y = CoordsUtil.scale(coords.y, dpr);
+        coords.width = CoordsUtil.scale(coords.width, dpr);
+        coords.height = CoordsUtil.scale(coords.height, dpr);
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/util/CoordsUtil.java:34]: do not change the input, create a new object `Coords`

File: vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/util/CoordsUtil.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+public final class CoordsUtil
+{
+    private CoordsUtil()
+    {
+    }
+
+    public static Coords scale(Coords coords, double dpr)
+    {
+        coords.x = CoordsUtil.scale(coords.x, dpr);
```
Comment (valfirst) [inline vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/screenshot/util/CoordsUtil.java:31]: Fully-qualified name is not needed:
`CoordsUtil.scale` -> `scale`

**Comment by EDbarvinsky** [comment]
[Last run](https://github.com/vividus-framework/vividus/runs/4711876825?check_suite_focus=true) related to changes in CI config

**Review by valfirst — APPROVED** [review]

**Pull request merged**